platform/engineering/design-docs/2020-05-26-content-validation.md

```diff
@@ -0,0 +1,208 @@
+
+##### Users
+
+To enable a faster publishing workflow, all tests (except link checking) are intentionally skipped during content-only deploys. That creates an opportunity for end users to encounter accessibility errors. 
+
+##### FE Tools team
+
+Invalid content causes warning logs during staging deploys, and breaks the build during production deploys. These broken builds then need manual intervention, and require coordinating with another team.
+
+##### Anybody that is trying to ship code for the VA.gov front end
+
+If invalid content is breaking the production build, and therefore blocking the production deploy, then no team on any contract is able to ship their frontend code to VA.gov.
+
+### High Level Design
+
+Once the `content-build` repo is successfully extracted from `vets-website`, the Metalsmith static content build pipeline will be deleted from `vets-website`. Webpack will handle the application build in `vets-website`, and Metalsmith will handle the static content build in the `content-build` repo. This will create two separate pipelines for built code to be deployed to S3, which will prevent failures in one from blocking the other.
+
+The validation of the static content coming from the CMS will occur during the Metalsmith build.
+
+## Specifics
+
+### Detailed Design
+
+The content validation will happen in a daily scheduled job.
+
+### Code Location
+
+The content validation script and scheduled job config will live in the [new `content-build` repo](https://github.com/department-of-veterans-affairs/content-build).
+
+### Testing Plan
+
+1. Configure a scheduled job to run a script once per workday
+2. Write an initial script that logs when the script is run
+3. Once the scheduled job is running at the correct time, replace the initial script with the actual script that runs the content validation
+
+### Logging
+
+Currently, invalid content is logged during staging deploys. And failed builds trigger generic [Slack notifications](https://dsva.slack.com/archives/C37M86Y8G/p1591099447374800).
+
+### Debugging
+
+##### Broken links
+
+1. Confirm that the content build is passing
+2. Modify an HTML file, like [`header.html`](https://github.com/department-of-veterans-affairs/content-build/blob/master/src/site/includes/header.html) to have a broken link
+3. Run the build again to confirm that it is failing
+
+##### Accessibility errors
+
+1. Confirm that the content build is passing
+2. Modify an HTML file, like [`header.html`](https://github.com/department-of-veterans-affairs/content-build/blob/master/src/site/includes/header.html) to have an accessibility error
+3. Run the build again to confirm that it is failing
+
+### Caveats
+
+- Our [accessibility check](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/assets/js/execute-axe-check.js#L8) only renders a banner on the preview server.
```

Comment:
This seems like it could easily get confused with the accessibility tests that Jenkins runs.  Can you differentiate them a bit more?

Reply:
I actually removed this caveat because I added it before I realized that there are actually two different types of a11y checks. 

1. During the Metalsmith build
2. During the Integration stage of Jenkins

Reply:
FYSA -

> During the Metalsmith build

This is totally non-blocking and runs only on the preview server. It _can_ be run during a regular build or watch task, but that's not what it was designed for.

> During the Integration stage of Jenkins

This _does_ block regular full deployments. The gist is that -

1. Everything is good and CI is passing
1. A new page is published in Drupal
1. CI runs on vets-website master or on a PR branch
1. The new page from Drupal is served into vets-website CI
1. The new page has an accessibility issue, and now all branches in vets-website have to fall back to the cached Drupal data on S3 in order to build successfully.
        - This doesn't always work. Sometimes the cache is considered invalid if the GraphQL queries have changed since the last cache was uploaded, so if there's an accessibility issue at the same time, CI is blocked for all branches.
1. Full deployments can't go out. 
    - Content-only deploys skip accessibility tests, so content can (ironically) still go out.
    - If vets-website master passed okay during the last build, we can do an app-code deployment by skipping the pre-release job in Jenkins. (Just unchecking the box for the deploy job)

So...we could technically achieve a full deployment by first doing an app-code deploy, then doing a content-only deploy. But we can't do them both at the same time if there are accessibility issues on a Drupal page.